renegade/types.py

```diff
@@ -0,0 +1,114 @@
+from pydantic import BaseModel, Field, ConfigDict, model_validator
+from enum import Enum
+from typing import Optional, Dict, Any
+from web3 import Web3
+from web3.types import TxParams
+import time
+
+class OrderSide(str, Enum):
+    BUY = "Buy"
+    SELL = "Sell"
+
+class BaseModelWithConfig(BaseModel):
+    """Base model with common configuration"""
+    model_config = ConfigDict(arbitrary_types_allowed=True, exclude_none=True)
+
+    def model_dump(self, **kwargs) -> Dict[str, Any]:
+        data = super().model_dump(**kwargs)
+        return self._remove_none_recursive(data)
+
+    def _remove_none_recursive(self, data: Any) -> Any:
+        if isinstance(data, dict):
+            return {
+                k: self._remove_none_recursive(v)
+                for k, v in data.items()
+                if v is not None
+            }
+        elif isinstance(data, list):
+            return [self._remove_none_recursive(item) for item in data]
+        return data
+
+class ApiExternalAssetTransfer(BaseModelWithConfig):
+    mint: str
+    amount: int
+
+class ApiTimestampedPrice(BaseModelWithConfig):
+    price: str
+    timestamp: int
+
+    @classmethod
+    def now(cls, price: float) -> "ApiTimestampedPrice":
+        return cls(
+            price=str(price),
+            timestamp=int(time.time() * 1000)
+        )
+
+class ApiExternalMatchResult(BaseModelWithConfig):
+    quote_mint: str
+    base_mint: str
+    quote_amount: int
+    base_amount: int
+    direction: OrderSide
+
+class FeeTake(BaseModelWithConfig):
+    relayer_fee: int
+    protocol_fee: int
+
+class ExternalOrder(BaseModelWithConfig):
+    quote_mint: str
+    base_mint: str
+    side: OrderSide
+    base_amount: Optional[int] = None
+    quote_amount: Optional[int] = None
+    min_fill_size: int
+
+    def model_post_init(self, __context) -> None:
+        self.quote_mint = Web3.to_checksum_address(self.quote_mint)
+        self.base_mint = Web3.to_checksum_address(self.base_mint)
```

Comment:
Why is it necessary to use checksummed addresses only for the `ExternalOrder` class?

Reply:
the web3 library requires them